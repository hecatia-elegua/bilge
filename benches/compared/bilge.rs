use bilge::{Bitsized, Number, FromBits, bitsize, u4, u12, u20, u5, u2, DebugBits};

#[inline(never)]
pub fn bilge(input: (u32, u32, u64, u16)) {
    let mut lpi = GicRedistributorLpi {
        control: RedistributorControl::from(input.0),
        implementer_identification: RedistributorImplementerIdentification::from(input.1),
        redistributor_type: RedistributorType::from(input.2),
    };
    assert_eq!(lpi.control.value, input.0);
    assert_eq!(lpi.implementer_identification.value, input.1);
    assert_eq!(lpi.redistributor_type.value, input.2);
    
    assert!(lpi.control.clear_enable_supported());
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(2054));
    lpi.implementer_identification.set_implementer_jep106(u12::new(input.3));
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(input.3));
    assert_eq!(lpi.redistributor_type.processor_number(), 63872);
}

#[derive(Debug)]
pub struct GicRedistributorLpi {
    control: RedistributorControl,
    implementer_identification: RedistributorImplementerIdentification,
    redistributor_type: RedistributorType,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorControl {
    //ro
    upstream_write_pending: bool,
    //zero
    reserved: u4,
    //or reserved
        disable_processor_selection_for_group_1_secure_interrupts: bool,
        disable_processor_selection_for_group_1_non_secure_interrupts: bool,
        disable_processor_selection_for_group_0_interrupts: bool,
    //zero
    reserved: u20,
    register_write_pending: bool,
    lpi_invalidate_registers_supported: bool,
    clear_enable_supported: bool,
    enable_lpis: bool,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorImplementerIdentification {
    //ro
    product_id: u8,
    //zero
    reserved: u4,
    //ro
    variant: u4,
    //ro
    revision: u4,
    implementer_jep106: u12,
}

#[bitsize(64)]
#[derive(DebugBits, FromBits)]
struct RedistributorType {
    affinity_value: u32,
    ppi_num: u5,
    virtual_sgi_supported: bool,
    common_lpi_affinity: u2, //TODO: enums everywhere
    processor_number: u16,
    resident_vpe_id: bool,
    mpam_supported: bool,
    control_cpgs_supported: bool,
    is_last: bool,
    direct_lpi_supported: bool,
    dirty: bool,
    virtual_lpi_supported: bool,
    physical_lpi_supported: bool,
}
