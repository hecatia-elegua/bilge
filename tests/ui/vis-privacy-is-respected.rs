use bilge::prelude::*;

mod hidden {
    use super::*;
    #[bitsize(96)]
    #[derive(FromBits)]
    pub struct Diary(u48, pub u48);

    #[bitsize(8)]
    #[derive(FromBits)]
    pub struct ArrayAt(pub [u2; 2], [u2; 2]);

    #[bitsize(128)]
    #[derive(FromBits)]
    pub struct Reserver {
        pub reserved: u8,
        pub(crate) padding: u8,
        reserved: u16,
        // the best use for padding
        pub padding: Diary,
    }
}
use hidden::*;

fn main() {
    let ascii = u48::new(0b01001001_00100000_01101100_01101111_01110110_01100101);
    let rusti = u48::new(0b00100000_01110010_01110101_01110011_01110100_00101110);
    let mut diary = Diary::new(ascii, rusti);
    diary.val_0();
    diary.val_1();
    diary.set_val_0(rusti);
    diary.set_val_1(ascii);

    let mut a = ArrayAt::from(42);
    a.val_0_at(1);
    a.set_val_0_at(1, u2::new(0));
    a.val_1_at(1);
    a.set_val_1_at(1, u2::new(0));

    // NOTE: as noted in another test case, `reserved` fields should mean "never used" and
    // might soon only be available to `DebugBits`, unless you need it. Then please open a github issue.
    // Right now though they have getters and do respect visibility:
    let mut r = Reserver::from(42);
    r.reserved_i();
    r.padding_i();
    r.reserved_ii();
    r.padding_ii();
}
