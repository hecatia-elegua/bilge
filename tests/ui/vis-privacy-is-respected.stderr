error[E0624]: associated function `val_0` is private
  --> tests/ui/vis-privacy-is-respected.rs:29:11
   |
5  |     #[bitsize(96)]
   |     -------------- private associated function defined here
...
29 |     diary.val_0();
   |           ^^^^^ private associated function

error[E0624]: associated function `set_val_0` is private
  --> tests/ui/vis-privacy-is-respected.rs:31:11
   |
5  |     #[bitsize(96)]
   |     -------------- private associated function defined here
...
31 |     diary.set_val_0(rusti);
   |           ^^^^^^^^^ private associated function

error[E0624]: associated function `val_1_at` is private
  --> tests/ui/vis-privacy-is-respected.rs:37:7
   |
9  |     #[bitsize(8)]
   |     ------------- private associated function defined here
...
37 |     a.val_1_at(1);
   |       ^^^^^^^^ private associated function

error[E0624]: associated function `set_val_1_at` is private
  --> tests/ui/vis-privacy-is-respected.rs:38:7
   |
9  |     #[bitsize(8)]
   |     ------------- private associated function defined here
...
38 |     a.set_val_1_at(1, u2::new(0));
   |       ^^^^^^^^^^^^ private associated function

error[E0624]: associated function `reserved_ii` is private
  --> tests/ui/vis-privacy-is-respected.rs:46:7
   |
13 |     #[bitsize(128)]
   |     --------------- private associated function defined here
...
46 |     r.reserved_ii();
   |       ^^^^^^^^^^^ private associated function
